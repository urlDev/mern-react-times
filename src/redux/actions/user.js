import axios from 'axios';
import toaster from 'toasted-notes';
import { push } from 'connected-react-router';

import NotificationComponent from '../../components/notification-component/NotificationComponent';

const url = process.env.REACT_APP_API_URL;

export const REGISTER_USER = 'REGISTER_USER';
export const LOGOUT_USER = 'LOGOUT_USER';
export const LOGIN_USER = 'LOGIN_USER';
export const DELETE_USER = 'DELETE_USER';
export const UPDATE_USER = 'UPDATE_USER';
export const GET_USER = 'GET_USER';
export const SET_TOKEN = 'SET_TOKEN';
export const USER_FETCH_ERROR = 'USER_FETCH_ERROR';
export const USER_LOADING = 'USER_LOADING';
export const USER_LOADING_END = 'USER_LOADING_END';
export const USER_MODAL_OPEN = 'USER_MODAL_OPEN';
export const USER_MODAL_CLOSE = 'USER_MODAL_CLOSE';
export const DELETE_MODAL_OPEN = 'DELETE_MODAL_OPEN';
export const DELETE_MODAL_CLOSE = 'DELETE_MODAL_CLOSE';
export const UPLOAD_AVATAR = 'UPLOAD_AVATAR';

export const registerUser = (user) => ({
  type: REGISTER_USER,
  payload: user,
});

export const logOutUser = () => ({
  type: LOGOUT_USER,
});

export const loginUser = (user) => ({
  type: LOGIN_USER,
  payload: user,
});

export const deleteUser = () => ({
  type: DELETE_USER,
});

export const updateUser = (user) => ({
  type: UPDATE_USER,
  payload: user,
});

export const getUser = (user) => ({
  type: GET_USER,
  payload: user,
});

export const setToken = (token) => ({
  type: SET_TOKEN,
  payload: token,
});

export const userFetchError = (error) => ({
  type: USER_FETCH_ERROR,
  payload: error,
});

export const userLoading = () => ({
  type: USER_LOADING,
});

export const userLoadingEnd = () => ({
  type: USER_LOADING_END,
});

export const userModalOpen = () => ({
  type: USER_MODAL_OPEN,
});

export const userModalClose = () => ({
  type: USER_MODAL_CLOSE,
});

export const deleteModalOpen = () => ({
  type: DELETE_MODAL_OPEN,
});

export const deleteModalClose = () => ({
  type: DELETE_MODAL_CLOSE,
});

export const uploadAvatar = (images) => ({
  type: UPLOAD_AVATAR,
  payload: images,
});

export const fetchRegisterUser = (user) => async (dispatch) => {
  dispatch(userLoading());
  try {
    const response = await axios.post(`${url}/profile/register`, user);
    const data = await response.data;
    return [
      dispatch(registerUser(data.user)),
      dispatch(userLoadingEnd()),
      dispatch(setToken(data.token)),
      localStorage.setItem('token', JSON.stringify(data.token)),
      localStorage.setItem('user', JSON.stringify(data.user)),
      dispatch(push('/')),
      toaster.notify(
        () => (
          <NotificationComponent
            success={true}
            text={`Welcome! Enjoy your time`}
          />
        ),
        { duration: 1500 }
      ),
    ];
  } catch (error) {
    return [
      dispatch(userFetchError(error)),
      dispatch(userLoadingEnd()),
      toaster.notify(
        () => (
          <NotificationComponent
            text={'Oops! Something went wrong!'}
            success={false}
          />
        ),
        { duration: 1500 }
      ),
    ];
  }
};

export const fetchLogoutUser = () => async (dispatch) => {
  const token = JSON.parse(localStorage.getItem('token'));

  const config = {
    headers: { Authorization: `Bearer ${token}` },
  };

  const path = window.location.pathname;

  try {
    await axios.post(`${url}/profile/logout`, null, config);

    return [
      path.includes('profile') && dispatch(push('/home')),
      dispatch(logOutUser()),
      toaster.notify(
        () => <NotificationComponent text={'Buh-Bye!'} success={true} />,
        { duration: 1500 }
      ),
    ];
  } catch (error) {
    return dispatch(userFetchError(error));
  }
};

export const fetchLoginUser = (user) => async (dispatch) => {
  dispatch(userLoading());
  try {
    const response = await axios.post(`${url}/profile/login`, user);
    const data = await response.data;
    return [
      dispatch(loginUser(data.user)),
      dispatch(userLoadingEnd()),
      dispatch(setToken(data.token)),
      localStorage.setItem('token', JSON.stringify(data.token)),
      localStorage.setItem('user', JSON.stringify(data.user)),
      dispatch(push('/home')),
      toaster.notify(
        () => (
          <NotificationComponent
            success={true}
            text={`Welcome! Enjoy your time`}
          />
        ),
        { duration: 1500 }
      ),
    ];
  } catch (error) {
    return [
      dispatch(userFetchError(error)),
      dispatch(userLoadingEnd()),
      toaster.notify(
        () => (
          <NotificationComponent
            text={'Oops! Something went wrong!'}
            success={false}
          />
        ),
        { duration: 1500 }
      ),
    ];
  }
};

// To get the user,
// but because I added an or statement in user and token initial states,
// app will check if there is user and token in localStorage
// so there will be no need for this.
// Adding just in case.
export const fetchUser = () => async (dispatch) => {
  const token = JSON.parse(localStorage.getItem('token'));

  const config = {
    headers: { Authorization: `Bearer ${token}` },
  };
  dispatch(userLoading());
  try {
    const response = await axios.get(`${url}/profile`, config);
    const data = await response.data;
    return [
      dispatch(registerUser(data.user)),
      dispatch(userLoadingEnd()),
      dispatch(setToken(data.token)),
      localStorage.setItem('token', JSON.stringify(data.token)),
      localStorage.setItem('user', JSON.stringify(data.user)),
    ];
  } catch (error) {
    return [dispatch(userFetchError(error)), dispatch(userLoadingEnd())];
  }
};

export const fetchUpdateUser = (user) => async (dispatch) => {
  const token = JSON.parse(localStorage.getItem('token'));

  const config = {
    headers: { Authorization: `Bearer ${token}` },
  };
  try {
    const response = await axios.patch(`${url}/profile`, user, config);
    const data = await response.data;
    return [
      dispatch(updateUser(data)),
      localStorage.setItem('user', JSON.stringify(data)),
      dispatch(push('/')),
      toaster.notify(
        () => (
          <NotificationComponent
            success={true}
            text={`Updated successfully!`}
          />
        ),
        { duration: 1500 }
      ),
    ];
  } catch (error) {
    return [
      dispatch(userFetchError(error)),
      toaster.notify(
        () => (
          <NotificationComponent
            text={'Oops! Something went wrong!'}
            success={false}
          />
        ),
        { duration: 1500 }
      ),
    ];
  }
};

export const fetchDeleteUser = () => async (dispatch) => {
  const token = JSON.parse(localStorage.getItem('token'));

  const config = {
    headers: { Authorization: `Bearer ${token}` },
  };
  try {
    await axios.delete(`${url}/profile`, config);
    return [
      localStorage.clear(),
      dispatch(push('/')),
      dispatch(deleteUser()),
      toaster.notify(
        () => <NotificationComponent success={true} text={`Buh-bye!`} />,
        { duration: 1500 }
      ),
    ];
  } catch (error) {
    return [
      dispatch(userFetchError(error)),
      toaster.notify(
        () => (
          <NotificationComponent
            success={false}
            text={`Something went wrong!`}
          />
        ),
        { duration: 1500 }
      ),
    ];
  }
};

export const fetchUploadAvatar = (input) => async (dispatch) => {
  const token = JSON.parse(localStorage.getItem('token'));

  const config = {
    headers: { Authorization: `Bearer ${token}` },
  };

  try {
    const response = await axios.post(`${url}/profile/avatar`, input, config);
    const data = await response.data;
    return [
      dispatch(uploadAvatar(data)),
      localStorage.setItem('user', JSON.stringify(data)),
      toaster.notify(
        () => (
          <NotificationComponent
            text={'Avatar is changed! Wow, that looks amazing! ;)'}
            success={true}
          />
        ),
        { duration: 1500 }
      ),
    ];
  } catch (error) {
    return [
      dispatch(userFetchError(error)),
      toaster.notify(
        () => (
          <NotificationComponent
            text={'Oops! Something went wrong!'}
            success={false}
          />
        ),
        { duration: 1500 }
      ),
    ];
  }
};
