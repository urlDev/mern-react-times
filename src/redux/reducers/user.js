import {
    REGISTER_USER,
    SET_TOKEN,
    USER_FETCH_ERROR,
    USER_LOADING,
    USER_LOADING_END,
} from '../actions/user';

const initialState = {
    loadingUser: false,
    errorUser: null,
    user: JSON.parse(localStorage.getItem('user')) || {},
    token: JSON.parse(localStorage.getItem('token')) || {},
};

const userReducer = (state = initialState, action) => {
    switch (action.payload) {
        case REGISTER_USER:
            return {
                ...state,
                user: action.payload,
            };
        case SET_TOKEN:
            return {
                ...state,
                token: action.payload,
            };
        case USER_FETCH_ERROR:
            return {
                ...state,
                error: action.payload,
            };
        case USER_LOADING:
            return {
                ...state,
                loading: true,
            };
        case USER_LOADING_END:
            return {
                ...state,
                loading: false,
            };
        default:
            return state;
    }
};

export default userReducer;