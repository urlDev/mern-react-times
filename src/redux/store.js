import { createStore, applyMiddleware } from 'redux';
import rootReducer from './rootReducer';

import thunk from 'redux-thunk';
import logger from 'redux-logger';

const middlewares = [thunk, logger];

export default createStore(rootReducer, applyMiddleware(...middlewares));