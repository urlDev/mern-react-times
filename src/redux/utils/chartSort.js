const chartSort = (prevState, newState) => {
    const tempState = [...prevState, [...newState]];

    let sortedState = [...tempState];

    sortedState = sortedState
        .sort((a, b) => a[1] - b[1])
        .map((stock) => stock.slice(0, 1));

    return sortedState;
};

export default chartSort;