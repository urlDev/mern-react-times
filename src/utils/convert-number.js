const convertNumber = (labelValue) => {
  // Nine Zeroes for Billions
  return Math.abs(Number(labelValue)) >= 1.0e9
    ? (Math.abs(Number(labelValue)) / 1.0e9).toFixed(2) + "B"
    : // Six Zeroes for Millions
    Math.abs(Number(labelValue)) >= 1.0e6
    ? (Math.abs(Number(labelValue)) / 1.0e6).toFixed(2) + "M"
    : // Three Zeroes for Thousands
    Math.abs(Number(labelValue)) >= 1.0e3
    ? (Math.abs(Number(labelValue)) / 1.0e3).toFixed(2) + "K"
    : Math.abs(Number(labelValue));
};

export default convertNumber;
