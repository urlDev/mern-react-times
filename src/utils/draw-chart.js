// Reusable function to draw historical price chart in different components
const drawChart = ({ width, reDraw, chartData, svg, d3, index }) => {
  // reDraw => Charts in market details
  // because they are drawn according to screen sizes

  let height;
  let padding;
  let chartWidth;
  let dates;

  // reDraw is for when marketDetails page is shown
  // Chart is constantly redrawn then so thats why I made a separation
  if (reDraw) {
    // changing chart width according to screen width
    chartWidth = width < 1400 ? (width < 700 ? width - 30 : width - 200) : 550;
    height = width < 1400 ? chartWidth * 0.65 : 345;
    padding = 50;
  } else {
    // when its used in market components, width and height
    chartWidth = 100;
    height = 70;
    padding = 10;
  }

  svg.attr("width", chartWidth).attr("height", height);

  if (reDraw) {
    dates = chartData.map((value) => new Date(value.date));
  } else {
    // In market components, there are more than one chart
    // So I get their indexes from the component
    dates = chartData[index].map((value) => new Date(value.date));
  }

  // xScale
  const xAxisScale = d3
    .scaleTime()
    .domain([d3.min(dates), d3.max(dates)])
    .range(
      // Depending if its reDrawn or width changes,
      // In smaller screens, I don't put any padding
      reDraw
        ? width < 500
          ? [0, chartWidth]
          : [padding, chartWidth - padding]
        : [padding, chartWidth - padding]
    );

  // yScale
  const yAxisScale = d3
    .scaleLinear()
    .domain([
      // Chart is drawn from close data on each dataset
      // So in here I am checking where the function is being used
      d3.min(reDraw ? chartData : chartData[index], (data) => data.close),
      d3.max(reDraw ? chartData : chartData[index], (data) => data.close),
    ])
    .range([height - padding, padding]);

  // Charts in market cards don't use axes
  if (reDraw) {
    //   create x Axes
    const xAxes = d3
      .axisBottom(xAxisScale)
      .ticks(width < 700 ? d3.timeDay.every(10) : 5)
      .tickFormat((date) => d3.timeFormat("%b %d, %I:%M")(date));

    // draw x axes
    width > 500 &&
      svg
        .append("g")
        .call(xAxes)
        .attr("transform", `translate(0,${height - padding})`);

    // create y axes
    const yAxes = d3.axisLeft(yAxisScale).tickFormat((d) => `$ ${d}`);

    // draw y axes
    width > 500 &&
      svg.append("g").call(yAxes).attr("transform", `translate(${padding},0)`);
  }

  // create line
  const line = d3
    .line()
    .x((d) => xAxisScale(new Date(d.date)))
    .y((d) => yAxisScale(d.close));

  // draw historical price path
  const path = svg
    .append("path")
    .data(reDraw ? [chartData] : [chartData[index]])
    .style("fill", " none")
    .attr("stroke", reDraw ? "black" : "white")
    .attr("stroke-width", reDraw ? "1.5" : "1.2")
    .attr("d", line);

  // get the length of the path for line transition
  const pathLength = path.node().getTotalLength();

  // transition for the line
  const transitionPath = d3.transition().ease(d3.easeSin).duration(2000);

  // setting pathsLength for strokes attrs and using transition
  // I made it minus pathLength because it was drawing it backwards
  path
    .attr("stroke-dashoffset", -pathLength)
    .attr("stroke-dasharray", pathLength)
    .transition(transitionPath)
    .attr("stroke-dashoffset", 0);

  path.interrupt("stroke-dashoffset");
  path.interrupt("stroke-dasharray");
};

export default drawChart;
