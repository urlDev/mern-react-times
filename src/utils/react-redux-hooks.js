// Mocking redux hooks
import {
  useSelector as originalUseSelector,
  useDispatch as originalUseDispatch,
} from "react-redux";

export const useSelector = (state) => originalUseSelector(state);
export const useDispatch = () => originalUseDispatch();
