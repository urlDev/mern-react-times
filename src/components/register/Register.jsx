import React from 'react';
import { Link, useHistory } from 'react-router-dom';
import { useDispatch } from 'react-redux';

import { fetchRegisterUser } from '../../redux/actions/user';

import { LoginContainer } from '../login/Login.styles';

const Register = () => {
  const initialState = {
    name: '',
    email: '',
    password: '',
  };
  // destructuring properties off of input
  const [{ name, email, password }, setInput] = React.useState(initialState);
  const dispatch = useDispatch();
  const history = useHistory();

  const handleChange = (e) => {
    // spread prevState and add new values with name, dynamically
    setInput((prevState) => ({
      ...prevState,
      [e.target.name]: e.target.value,
    }));
  };

  const handleSubmit = async (e) => {
    e.preventDefault();

    await dispatch(fetchRegisterUser({ name, email, password }));
    // set state back to initial
    setInput({ ...initialState });
    history.push('/');
  };

  return (
    <LoginContainer>
      <form onSubmit={handleSubmit}>
        <input
          type="text"
          name="name"
          aria-label="name"
          placeholder="Name"
          onChange={handleChange}
          value={name}
        />

        <input
          type="email"
          name="email"
          aria-label="email"
          placeholder="Email"
          onChange={handleChange}
          value={email}
        />
        <input
          type="password"
          name="password"
          aria-label="password"
          placeholder="Password"
          onChange={handleChange}
          value={password}
        />
        <button aria-label="login" type="submit">
          Register
        </button>
      </form>
      <h1 style={{ textAlign: 'center', color: 'white', marginBottom: '40px' }}>
        Do you have an account?{' '}
        <Link to="/profile/login" style={{ color: 'lightgray' }}>
          Login
        </Link>
      </h1>
    </LoginContainer>
  );
};

export default Register;
