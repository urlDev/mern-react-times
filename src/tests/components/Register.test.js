import React from "react";
import { shallow } from "enzyme";
import moxios from "moxios";

import * as ReactReduxHooks from "utils/react-redux-hooks";
import { mockStore } from "tests/store";
import Register from "components/register/Register";
import { data, userInputRegister } from "tests/fixtures/user";
import {
  fetchRegisterUser,
  REGISTER_USER,
  SET_TOKEN,
  USER_LOADING,
  USER_LOADING_END,
} from "redux/actions/user";

describe("Testing Register component", () => {
  let store;
  let wrapper;

  beforeEach(() => {
    store = mockStore({});

    jest
      .spyOn(ReactReduxHooks, "useDispatch")
      .mockImplementation(() => store.dispatch);

    jest
      .spyOn(ReactReduxHooks, "useSelector")
      .mockImplementation((state) => store.getState());

    wrapper = shallow(<Register store={store} />);

    moxios.install();
  });

  afterEach(() => {
    moxios.uninstall();
  });
  test("Should render component", () => {
    expect(wrapper).toHaveLength(1);
    expect(wrapper.find("button").text()).toEqual("Register");
  });
  test("Should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("Should change the input value", () => {
    const value = "Can";

    wrapper
      .find("input")
      .first()
      .simulate("change", { target: { value, name: "name" } });

    expect(wrapper.find("input").first().prop("value")).toEqual(value);
  });
  test("Should submit the form when user registers", async () => {
    wrapper.find("form").simulate("submit", { preventDefault: () => {} });

    moxios.wait(() => {
      const request = moxios.requests.mostRecent();
      request.respondWith({
        status: 200,
        response: data,
      });
    });

    const expectedActions = [
      { type: USER_LOADING },
      { type: USER_LOADING },
      {
        type: REGISTER_USER,
        payload: data.user,
      },
      { type: USER_LOADING_END },
      { type: SET_TOKEN, payload: data.token },
      {
        type: "@@router/CALL_HISTORY_METHOD",
        payload: { method: "push", args: ["/"] },
      },
    ];

    await store.dispatch(fetchRegisterUser(userInputRegister));

    const actions = store.getActions();
    expect(actions).toEqual(expectedActions);
  });
});
