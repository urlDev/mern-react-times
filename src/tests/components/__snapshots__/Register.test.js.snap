// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Register component Should match snapshot 1`] = `
<styled.div>
  <form
    onSubmit={[Function]}
  >
    <input
      aria-label="name"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <input
      aria-label="email"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="email"
      value=""
    />
    <input
      aria-label="password"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
    <button
      aria-label="login"
      type="submit"
    >
      Register
    </button>
  </form>
  <h1
    style={
      Object {
        "color": "white",
        "marginBottom": "40px",
        "textAlign": "center",
      }
    }
  >
    Do you have an account?
     
    <Link
      style={
        Object {
          "color": "lightgray",
        }
      }
      to="/profile/login"
    >
      Login
    </Link>
  </h1>
</styled.div>
`;
